active/0000-stateful-helpers.md

```diff
@@ -0,0 +1,322 @@
+- Start Date: 2015-05-17
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Stateful helpers are a class-based way to define helpers. Stateful helpers:
+
+  * Have a single return value
+  * Can store and read state
+  * Have lifecycle hooks analagous to components where appropriate. For
+    example, a stateful helper may call `recompute` at any time to generate a new
+    value.
+  * Are a superset of traditional function-based helpers. They can do more,
+    but in many cases a traditional helper is appropriate.
+
+They fill a gap in Ember's current template APIs:
+
+|                  | has positional params | has layout | can yield template | has lifecycle, instance | can control rerender |
+|------------------|-----------------------|------------|--------------------|-------------------------|----------------------|
+| components       | Yes                   | Yes        | Yes                | Yes                     | Yes                  |
+| stateful helpers | Yes                   | No         | Yes                | Yes                     | Yes                  |
+| function helpers | Yes                   | No         | Yes                | No                      | No                   |
+
+Example usage:
+
+```js
+// app/helpers/full-name.js
+import Ember from "ember";
+
+export default Ember.Helper.extend({
+  nameBuilder: Ember.inject.service(),
+  compute(params) {
+    const builder = this.get('nameBuilder');
+    return builder.fullName(params[0], params[1]);
+  }
+});
+```
+
+Use an stateful helper anywhere a traditional helper is valid:
+
+```hbs
+{{full-name 'Bigtime' 'Beagle'}}
+{{input value=(full-name 'Gyro' 'Gearloose') readonly=true}}
+```
+
+# Motivation
+
+Helpers in Ember are pure functions. This make them easy to reason about, but
+also overly simplistic. It is difficult to write a helper that recomputes due to
```

Comment:
I think I would say this a little bit differently.

Historically, helpers in Ember have had access to a lot of ambient information in the scope. This made them powerful, but also error-prone. In practice, they were used by addon authors to implement powerful functionality.

Glimmer does not want to expose the guts of the scoping system to user code, so our initial stab at helpers was to make them pure functions: data in, data out. That works well for a lot of cases, and is very ergonomic, but is a regression for addons that need the value they emit to be based on external state. This proposal is an attempt to recoup enough expressiveness so that these addons can retool for Glimmer without exposing the guts of the scoping system to addons again.